File: src/core/Cryptos/PositionInstance.cs
using System;

namespace core.Cryptos
{
    // PositionInstance models a stock position from the time the first share is opened
    // to the time when the last share is sold.
    public class PositionInstance
    {
        private DateTimeOffset? _firstOpen = null;
        private decimal _quantity = 0;

        public PositionInstance(string token)
        {
            Token = token;
        }

        public int DaysHeld => _firstOpen != null ? (int)((!IsClosed ? DateTimeOffset.UtcNow : Closed.Value).Subtract(_firstOpen.Value)).TotalDays : 0;
        public decimal Cost { get; private set; } = 0;
        public decimal Return { get; private set; } = 0;
        public decimal Percentage => Cost == 0 ? 0 : Math.Round((Return - Cost) / Cost, 4);
        public decimal Profit => Return - Cost;
        public bool IsClosed => Closed != null;
        public string Token { get; }
        public DateTimeOffset? Closed { get; private set; }

        public void Buy(decimal quantity, decimal dollarAmountSpent, DateTimeOffset when)
        {
            if (_quantity == 0)
            {
                _firstOpen = when;
            }

            _quantity += quantity;

            Cost += dollarAmountSpent;
        }

        public void Sell(decimal quantity, decimal dollarAmountReceived, DateTimeOffset when)
        {
            _quantity -= quantity;

            if (_quantity < 0)
            {
                throw new InvalidOperationException("Transaction would make amount owned invalid");
            }

            if (_quantity == 0)
            {
                Closed = when;
            }

            Return += dollarAmountReceived;
        }
    }
}

File: src/core2.fs/Cryptos/PositionInstance.fs
namespace core.Cryptos

open System

[<AllowNullLiteral>]
type PositionInstance(token: string) =
    let mutable _firstOpen: DateTimeOffset option = None
    let mutable _quantity = 0m
    let mutable _cost = 0m
    let mutable _return = 0m
    let mutable _closed: DateTimeOffset option = None

    member this.DaysHeld =
        match _firstOpen with
        | Some fo -> int ((match _closed with Some c -> c | None -> DateTimeOffset.UtcNow).Subtract(fo).TotalDays)
        | None -> 0

    member _.Cost = _cost
    member _.Return = _return
    member this.Percentage = if _cost = 0m then 0m else Math.Round((_return - _cost) / _cost, 4)
    member this.Profit = _return - _cost
    member _.IsClosed = _closed.IsSome
    member _.Token = token
    member _.Closed = _closed

    member this.Buy(quantity: decimal, dollarAmountSpent: decimal, ``when``: DateTimeOffset) =
        if _quantity = 0m then
            _firstOpen <- Some ``when``
        _quantity <- _quantity + quantity
        _cost <- _cost + dollarAmountSpent

    member this.Sell(quantity: decimal, dollarAmountReceived: decimal, ``when``: DateTimeOffset) =
        _quantity <- _quantity - quantity
        if _quantity < 0m then
            raise (InvalidOperationException("Transaction would make amount owned invalid"))
        if _quantity = 0m then
            _closed <- Some ``when``
        _return <- _return + dollarAmountReceived
